client/x/evmstaking/keeper/withdraw.go

```diff
@@ -82,86 +88,103 @@ func (k Keeper) ExpectedPartialWithdrawals(ctx context.Context) ([]estypes.Withd
 		if err != nil {
 			return nil, errors.Wrap(err, "get validator delegations")
 		}
-		swept += uint32(len(delegators))
-		log.Debug(
-			ctx, "Get all delegators of validator",
-			"val_addr", valAddr.String(),
-			"delegator_amount", len(delegators),
-		)
-		// Get delegator rewards.
-		for i := range delegators {
+
+		nextDelegators := delegators[nextValDelIndex:]
+
+		// Check if the sweep should stop prematurely as the current delegator loop exceeds the sweep bound while sweeping.
+		shouldStopPrematurely := swept+uint32(len(nextDelegators)) > sweepBound
+		stoppedPrematurely := false
+
+		// Iterate on the validator's delegator rewards in the range [nextValDelIndex, len(delegators)].
+		for i := range nextDelegators {
 			// Get end current period and calculate rewards.
 			endingPeriod, err := k.distributionKeeper.IncrementValidatorPeriod(ctx, validatorSet[nextValIndex])
 			if err != nil {
 				return nil, err
 			}
-			delRewards, err := k.distributionKeeper.CalculateDelegationRewards(ctx, validatorSet[nextValIndex], delegators[i], endingPeriod)
+
+			delRewards, err := k.distributionKeeper.CalculateDelegationRewards(ctx, validatorSet[nextValIndex], nextDelegators[i], endingPeriod)
 			if err != nil {
 				return nil, err
 			}
-			if delegators[i].DelegatorAddress == valAccAddr.String() {
+
+			if nextDelegators[i].DelegatorAddress == valAccAddr.String() {
 				delRewards = delRewards.Add(valCommission.Commission...)
 			}
 			delRewardsTruncated, _ := delRewards.TruncateDecimal()
 			bondDenomAmount := delRewardsTruncated.AmountOf(sdk.DefaultBondDenom).Uint64()
 
-			log.Debug(
-				ctx, "Calculate delegator rewards",
-				"val_addr", valAddr.String(),
-				"del_addr", delegators[i].DelegatorAddress,
-				"rewards_amount", bondDenomAmount,
-				"ending_period", endingPeriod,
-			)
-
 			if bondDenomAmount >= minPartialWithdrawalAmount {
-				delEvmAddr, err := k.DelegatorMap.Get(ctx, delegators[i].DelegatorAddress)
+				delEvmAddr, err := k.DelegatorMap.Get(ctx, nextDelegators[i].DelegatorAddress)
 				if err != nil {
 					return nil, errors.Wrap(err, "map delegator pubkey to evm address")
 				}
+
 				withdrawals = append(withdrawals, estypes.NewWithdrawal(
 					uint64(sdk.UnwrapSDKContext(ctx).BlockHeight()),
-					delegators[i].DelegatorAddress,
+					nextDelegators[i].DelegatorAddress,
 					valAddr.String(),
 					delEvmAddr,
 					bondDenomAmount,
 				))
+			}
 
-				log.Debug(
-					ctx, "Found an eligible partial withdrawal",
-					"val_addr", valAddr.String(),
-					"del_addr", delegators[i].DelegatorAddress,
-					"del_evm_addr", delEvmAddr,
-					"rewards_amount", bondDenomAmount,
-				)
+			nextValDelIndex++
+
+			// Current delegator loop exceeds the sweep bound while sweeping, so we break prematurely from validator delegation sweep.
+			// The current value of `nextValIndex` and `nextValDelIndex` will be used in the next sweep (next block).
+			if shouldStopPrematurely && swept+uint32(i) > sweepBound {
```

Comment:
Why do we need shouldStopPrematurely?

Reply:
The correct boundary should be swept + uint32(I) >= sweepBound and it should be checked at the beginning of the loop. You can use swept == 1024 and I == 0 as a test example to validate.

Reply:
Logic changed in 13e0724 to loop all delegations of a validator if the length <= the remaining sweep amount, otherwise loop only the first N delegations of the validator and break.